clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java

```diff
@@ -63,64 +65,71 @@
+    public void subscribe(List<String> topics, RebalanceListener listener) {
+        if (!this.userAssignment.isEmpty())
             throw new IllegalStateException("Subscription to topics and partitions are mutually exclusive");
-        if (!this.subscribedTopics.contains(topic)) {
-            this.subscribedTopics.add(topic);
+
+        if (this.listener == null || !this.listener.equals(listener))
+            this.listener = listener;
+
+        if (!this.subscription.equals(new HashSet<>(topics))) {
+            this.subscription.clear();
+            this.subscription.addAll(topics);
             this.needsPartitionAssignment = true;
-        }
-    }
 
-    public void unsubscribe(String topic) {
-        if (!this.subscribedTopics.contains(topic))
-            throw new IllegalStateException("Topic " + topic + " was never subscribed to.");
-        this.subscribedTopics.remove(topic);
-        this.needsPartitionAssignment = true;
-        final List<TopicPartition> existingAssignedPartitions = new ArrayList<>(assignedPartitions());
-        for (TopicPartition tp: existingAssignedPartitions)
-            if (topic.equals(tp.topic()))
-                clearPartition(tp);
+            // Remove any assigned partitions which are no longer subscribed to
+            Iterator<Map.Entry<TopicPartition, TopicPartitionState>> iterator = assignment.entrySet().iterator();
+            while (iterator.hasNext()) {
+                Map.Entry<TopicPartition, TopicPartitionState> entry = iterator.next();
+                if (!subscription.contains(entry.getKey().topic()))
+                    iterator.remove();
+            }
+        }
     }
 
     public void needReassignment() {
         this.needsPartitionAssignment = true;
     }
 
-    public void subscribe(TopicPartition tp) {
-        if (!this.subscribedTopics.isEmpty())
+    public void assign(List<TopicPartition> partitions) {
+        if (!this.subscription.isEmpty())
             throw new IllegalStateException("Subscription to topics and partitions are mutually exclusive");
-        this.subscribedPartitions.add(tp);
-        addAssignedPartition(tp);
-    }
 
-    public void unsubscribe(TopicPartition partition) {
-        if (!subscribedPartitions.contains(partition))
-            throw new IllegalStateException("Partition " + partition + " was never subscribed to.");
-        subscribedPartitions.remove(partition);
-        clearPartition(partition);
-    }
-    
-    private void clearPartition(TopicPartition tp) {
-        this.assignedPartitions.remove(tp);
+        this.userAssignment.clear();
+        this.userAssignment.addAll(partitions);
+
+        for (TopicPartition partition : partitions)
+            if (!assignment.containsKey(partition))
+                addAssignedPartition(partition);
+
+        // FIXME: This can probably be done more efficiently
+        Iterator<Map.Entry<TopicPartition, TopicPartitionState>> iterator = assignment.entrySet().iterator();
```

Comment:
Again, ines L#118 to L#123 can be replaced by:
```
assigment.keySet().retainAll(userAssignment);
```
same effect, as we want the intersection of ``assigment`` and ``userAssigment``